File: 2414100208-黄楚民/第八周作业/app/src/main/java/com/example/news2/MainActivity.java
package com.example.news2;

import android.os.Bundle;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.example.news2.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private NewsViewModel viewModel;
    private NewsAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        // 初始化 ViewModel
        viewModel = new ViewModelProvider(this).get(NewsViewModel.class);

        // 初始化 RecyclerView
        binding.rvNews.setLayoutManager(new LinearLayoutManager(this));
        adapter = new NewsAdapter();
        binding.rvNews.setAdapter(adapter);

        // 观察 LiveData
        viewModel.getNewsLiveData().observe(this, articles -> {
            adapter.updateData(articles);
        });

        viewModel.getErrorLiveData().observe(this, error -> {
            Toast.makeText(this, error, Toast.LENGTH_SHORT).show();
        });

        // 刷新按钮
        binding.btnRefresh.setOnClickListener(v -> viewModel.refreshNews());

        // 首次进入页面自动加载；旋转屏幕后 ViewModel 保留，不会重复请求。
        viewModel.loadNewsIfNeeded();
    }
}

File: 2414100208-黄楚民/第十一周作业/app/src/main/java/com/example/downloader/MainActivity.kt
package com.example.downloader

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.downloader.databinding.ActivityMainBinding
import java.io.File

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: DownloadViewModel
    private lateinit var adapter: ImageAdapter
    private val addedFiles = mutableSetOf<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this)[DownloadViewModel::class.java]

        // 初始化 RecyclerView
        adapter = ImageAdapter()
        binding.rvImages.layoutManager = LinearLayoutManager(this)
        binding.rvImages.adapter = adapter

        // 观察下载状态
        viewModel.state.observe(this) { state ->
            binding.tvStatus.text = state.status
            binding.btnStart.isEnabled = !state.isDownloading
            binding.btnCancel.isEnabled = state.isDownloading

            if (state.total > 0) {
                val percent = (state.progress * 100) / state.total
                binding.progressBar.progress = percent
                binding.tvProgress.text = "${state.progress}/${state.total} ($percent%)"
            }

            // 新图片下载完成时添加到列表（避免重复）
            state.completedFile?.let { file ->
                if (file.exists() && addedFiles.add(file.absolutePath)) {
                    adapter.addImage(file)
                }
            }
        }

        // 开始下载
        binding.btnStart.setOnClickListener {
            adapter.clear()
            addedFiles.clear()
            viewModel.startDownload()
        }

        // 取消下载
        binding.btnCancel.setOnClickListener {
            viewModel.cancelDownload()
        }
    }
}
